llvm/utils/TableGen/DynamicTableEmitter.cpp

```diff
@@ -0,0 +1,272 @@
+      ->getValue();
+}
+
+struct GenericField {
+  std::string Name;
+  RecTy *RecType = nullptr;
+  bool IsCode = false;
+  bool IsList = false;
+
+  GenericField(StringRef Name) : Name(std::string(Name)) {}
+};
+
+struct DynamicTable {
+  std::string Name;
+  ArrayRef<SMLoc> Locs; // Source locations from the Record instance.
+  std::string PreprocessorGuard;
+  std::string CppTypeName;
+  SmallVector<GenericField, 2> Fields;
+  std::vector<Record *> Entries;
+
+  const GenericField *getFieldByName(StringRef Name) const {
+    for (const auto &Field : Fields) {
+      if (Name == Field.Name)
+        return &Field;
+    }
+    return nullptr;
+  }
+};
+
+class DynamicTableEmitter {
+  RecordKeeper &Records;
+  std::set<std::string> PreprocessorGuards;
+
+public:
+  DynamicTableEmitter(RecordKeeper &R) : Records(R) {}
+
+  void run(raw_ostream &OS);
+
+private:
+  std::string primaryRepresentation(SMLoc Loc, const GenericField &Field,
+                                    Init *I) {
+    if (StringInit *SI = dyn_cast<StringInit>(I)) {
+      if (Field.IsCode || SI->hasCodeFormat())
+        return std::string(SI->getValue());
+
+      return SI->getAsString();
+    } else if (BitsInit *BI = dyn_cast<BitsInit>(I))
+      return "0x" + utohexstr(getAsInt(BI));
+    else if (BitInit *BI = dyn_cast<BitInit>(I))
+      return BI->getValue() ? "true" : "false";
+    else if (Field.IsList) {
+      if (auto LI = dyn_cast<ListInit>(I)) {
+        std::stringstream res;
+        auto values = LI->getValues();
+        bool IsAspect = (Field.Name == "aspects");
+        for (const auto &[idx, val] : enumerate(values)) {
+          if (idx > 0)
+            res << ", ";
+
+          if (IsAspect) {
+            auto rec = LI->getElementAsRecord(idx);
+            res << "\"" << rec->getValueAsString("Name").str() << "\"";
+          } else
+            res << val->getAsString();
+          }
+          return res.str();
+      }
+      PrintFatalError(Loc,
+                      Twine("Entry for field '") + Field.Name + "' is null");
+    }
+    PrintFatalError(Loc, Twine("invalid field type for field '") + Field.Name + 
+                             "'; expected: bit, bits, string, or code");
+  }
+
+  void emitDynamicTable(const DynamicTable &Table, raw_ostream &OS);
+  void emitIfdef(StringRef Guard, raw_ostream &OS);
+
+  bool parseFieldType(GenericField &Field, Init *II);
+  void collectTableEntries(DynamicTable &Table,
+                           const std::vector<Record *> &Items);
+};
+
+} // End anonymous namespace.
+
+void DynamicTableEmitter::emitIfdef(StringRef Guard, raw_ostream &OS) {
+  OS << "#ifdef " << Guard << "\n";
+  PreprocessorGuards.insert(std::string(Guard));
+}
+
+void DynamicTableEmitter::emitDynamicTable(const DynamicTable &Table,
+                                           raw_ostream &OS) {
+  emitIfdef((Twine("GET_") + Table.PreprocessorGuard + "_IMPL").str(), OS);
```

Comment:
This is a minor comment since there is only one use of `emitIfdef`, but it would probably be better to accept `Twine` in the method instead of `StringRef`, which would allow to make call to `.str()` part of the method itself.

Note that currently there is a bunch of conversions happening: `Twine` -> `std::string` -> `StringRef` -> `std::string`. Most of them are no-ops if I understand correctly, but still, types in callee and caller are a bit incosistent.